File: src/studio/program/dictionary/def/DiscreteInput.java
package studio.program.dictionary.def;

import studio.program.Var;
import studio.program.dictionary.BlockDefinition;
import studio.program.dictionary.PinDefinition;

public class DiscreteInput extends BlockDefinition {
    public static final String NAME = "discrete_input";

    public DiscreteInput() {
        super(NAME);

        pins.add(0, new PinDefinition(
                PinDefinition.Flow.OUTPUT,
                PinDefinition.Side.RIGHT,
                Var.Type.DISCRETE_SIGNAL
        ));
    }
}

File: src/studio/blocklib/definitions/DiscreteInput.kt
package studio.blocklib.definitions

import studio.blocklib.BlockDefinition
import studio.program.Block
import studio.program.Pin
import studio.program.Variable
import studio.shape.Circle

class DiscreteInput : BlockDefinition() {
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    // CONSTANTS
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    companion object {
        val NAME: String = "discrete_input"
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    // CONSTRUCTOR
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    init {
        genFn = """

        @PIN(0) = digitalRead(@VAR(pin));

        """.trimIndent()

        genInit = """

        pinMode(@VAR(pin), INPUT);

        """.trimIndent()
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    // PUBLIC FUNCTIONS
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    override fun constructInstance(): Block {
        val ins: Block = Block(NAME)

        ins.body = Circle()
        ins.symbol = "m -20 15 l 7 0 l 0 -30 l 26 0 l 0 30 l 7 0";

        ins.pins.add(Pin.create()
                .withParent(ins)
                .withVariable(Variable(Variable.Type.DISCRETE_SIGNAL))
                .atIndex(0)
                .ofFlow(Pin.Flow.OUTPUT)
                .onSide(Pin.Side.RIGHT)
                .attachedAt(40, 0))

        ins.vars.put("pin", Variable(Variable.Type.U8, 5));

        return ins
    }
}